src/content/reference/react-dom/hooks/useFormState.md

```diff
@@ -0,0 +1,291 @@
+
+</Canary>
+
+<Intro>
+
+`useFormState` is a Hook that allows you to read the return value of the form action after a form is submitted.
+
+```js
+const [state, formAction] = useFormState(action, initalState);
+```
+
+</Intro>
+
+<InlineToc />
+
+---
+
+## Reference {/*reference*/}
+
+### `useFormState()` {/*useformstate*/}
+
+In the context of React Server Components, an *action* is a function that may be [executed when a form is submitted](/reference/react-dom/components/form). You can execute actions on the server or on the client.
+
+{/* TODO T164397693: link to actions documentation once it exists */}
+
+Call `useFormState` at the top level of your component to see the return value of an action after submitting a form. You pass `useFormState` an existing action as well as an initial state, and it returns a new action that you use when submitting your form, along with the latest form state.
+
+```js
+function AddToCart({itemID}) {
+  const [message, formAction] = useFormState(addToCartAction, null);
+  return (
+    <form action={formAction}>
+      <input type="hidden" name="itemID" value={itemID} />
+      <button type="submit" label="Add to cart" />
+      <p>
+        {message}
+      </p>
+    </form>
+  )
+}
+```
+
+The form state is the value returned by the action when the form was last submitted. If the form has not yet been submitted, it is the initial state that you pass.
+
+If used with a server action, `useFormState` allows the server's response from submitting the form to be shown even before hydration has completed.
+
+[See more examples below.](#usage)
+
+#### Parameters {/*parameters*/}
+
+* `action`: The action to be performed when the form is submitted. When the action is called, it will receive the previous state of the form (initially the `initialState` that you pass, subsequently its previous return value) as its initial argument, followed by the arguments that an action normally receives.
+* `initialState`: The value you want the state to be initially. It can be any serializable value. This argument is ignored after the form is first submitted.
+
+{/* TODO T164397693: link to serializable values docs once it exists */}
+
+
+#### Returns {/*returns*/}
+
+`useFormState` returns an array with exactly two values:
+
+1. The current state. During the first render, it will match the `initialState` you have passed. After the form is submitted, it will match the value returned by the action.
+2. A new action that you can pass as the `action` prop to your `form` component.
+
+#### Caveats {/*caveats*/}
+
+* When used with a framework that supports React Server Components, `useFormState` lets you make forms interactive before JavaScript has executed on the client. When used without Server Components, there is no advantage to using it over component local state.
```

Comment:
As someone who is new to understanding `useFormState`, this caveat helped me understand why I would want to use this and the benefits. Perhaps we could incorporate the benefits that it lets you make forms interactive before JavaScript has executed on the client earlier

Reply:
I'm not sure I would say `no advantage` as that sounds like you're pushing people to useState (whereas I could imagine some people preferring useFormState's API), but you could say it is equivalent.